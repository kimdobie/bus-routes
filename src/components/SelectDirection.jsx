import React from 'react';
import PropTypes from 'prop-types';
import Form from 'react-bootstrap/Form';

const axios = require('axios');

class SelectDirection extends React.Component {
  constructor(props) {
    super(props);
    const { direction, route } = this.props;
    this.state = {
      direction, // id of a route
      directions: [], // array of route objects
      route
    };
    this.getDirections = this.getDirections.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.getDirections();
  }

  getDirections() {
    const { route } = this.state;
    if (route === '') {
      this.setState({ directions: [] });
      return;
    }
    axios
      .get(`/api/Directions/${route}`)
      .then(directions => {
        this.setState({ directions: directions.data });
      })
      .catch(error => {
        // eslint-disable-next-line no-console
        console.log('There was an issue getting directions:', error);
      });
  }

  setOptions() {
    const { directions } = this.state;
    return directions.map(directionObj => {
      return (
        <option value={directionObj.Value} key={directionObj.Value}>
          {directionObj.Text}
        </option>
      );
    });
  }

  handleChange(event) {
    const { setDirection } = this.props;
    const newDirection = event.target.value;
    this.setState({ direction: newDirection });
    setDirection(newDirection);
  }

  render() {
    const { direction } = this.state;
    return (
      <Form>
        <Form.Group controlId='formDirection'>
          <Form.Label>Direction:</Form.Label>
          <Form.Control
            as='select'
            value={direction}
            onChange={this.handleChange}
          >
            <option value=''>Select a direction</option>
            {this.setOptions()}
          </Form.Control>
        </Form.Group>
      </Form>
    );
  }
}
export default SelectDirection;

SelectDirection.propTypes = {
  direction: PropTypes.string,
  route: PropTypes.string,
  setDirection: PropTypes.func
};
SelectDirection.defaultProps = {
  direction: '',
  route: '',
  setDirection: () => {}
};
