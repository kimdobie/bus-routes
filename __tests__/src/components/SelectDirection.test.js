import React from 'react';
import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import SelectDirection from '../../../src/components/SelectDirection';

const fs = require('fs');
const path = require('path');
const axios = require('axios');

const { axe, toHaveNoViolations } = require('jest-axe');

expect.extend(toHaveNoViolations);

configure({ adapter: new Adapter() });

jest.mock('axios');

describe('Select Direction tests', () => {
  const routes = fs.readFileSync(
    path.join(__dirname, '../../../__fixtures__/routes.json')
  );
  const directions = fs.readFileSync(
    path.join(__dirname, '../../../__fixtures__/directions.json')
  );
  const stops = fs.readFileSync(
    path.join(__dirname, '../../../__fixtures__/stops.json')
  );
  axios.setRoutes(JSON.parse(routes));
  axios.setDirections(JSON.parse(directions));
  axios.setStops(JSON.parse(stops));

  test('Snapshot test without params', async () => {
    // eslint-disable-next-line react/jsx-filename-extension
    const selectDirection = await shallow(<SelectDirection />);
    expect(selectDirection.find('option')).toHaveLength(1);
    expect(selectDirection).toMatchSnapshot();
  });
  test('Snapshot test with params', async () => {
    const selectDirection = await shallow(
      <SelectDirection route='123' direction='456' />
    );
    expect(selectDirection.find('option')).toHaveLength(
      JSON.parse(directions).length + 1
    );
    expect(selectDirection).toMatchSnapshot();
  });
  test('Accessibility check', async () => {
    const wrapper = await mount(
      <SelectDirection route='123' direction='456' />
    );
    const results = await axe(wrapper.getDOMNode());
    // console.log('Axe violations', results.violations);
    expect(results).toHaveNoViolations();
  });

  test('Select direction', async () => {
    const selectDirection = await shallow(<SelectDirection route='123' />);
    expect(selectDirection.state().direction).toEqual('');
    expect(selectDirection.find('FormControl').prop('value')).toEqual('');
    const mockEvent = {
      target: {
        value: '789'
      }
    };
    selectDirection.instance().handleChange(mockEvent);
    expect(selectDirection.state().direction).toEqual('789');
    expect(selectDirection.find('FormControl').prop('value')).toEqual('789');
  });

  test('Change direction to ""', async () => {
    const selectDirection = await shallow(
      <SelectDirection route='123' direction='789' />
    );
    expect(selectDirection.state().direction).toEqual('789');
    expect(selectDirection.state().directions).toHaveLength(
      JSON.parse(directions).length
    );
    expect(selectDirection.find('FormControl').prop('value')).toEqual('789');

    const mockEvent = {
      target: {
        value: ''
      }
    };
    selectDirection.instance().handleChange(mockEvent);
    expect(selectDirection.state().direction).toEqual('');
    expect(selectDirection.find('FormControl').prop('value')).toEqual('');
  });
  test('Direction set but route missing from props', async () => {
    const selectDirection = await shallow(<SelectDirection direction='456' />);
    expect(selectDirection.state().directions).toEqual([]);
  });
});
