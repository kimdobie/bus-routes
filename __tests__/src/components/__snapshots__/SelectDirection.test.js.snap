// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Direction tests Snapshot test with params 1`] = `
<Form
  inline={false}
>
  <FormGroup
    controlId="formDirection"
  >
    <FormLabel
      column={false}
      srOnly={false}
    >
      Direction:
    </FormLabel>
    <FormControl
      as="select"
      onChange={[Function]}
      value="456"
    >
      <option
        value=""
      >
        Select a direction
      </option>
      <option
        key="2"
        value="2"
      >
        EASTBOUND
      </option>
      <option
        key="3"
        value="3"
      >
        WESTBOUND
      </option>
    </FormControl>
  </FormGroup>
</Form>
`;

exports[`Select Direction tests Snapshot test without params 1`] = `
<Form
  inline={false}
>
  <FormGroup
    controlId="formDirection"
  >
    <FormLabel
      column={false}
      srOnly={false}
    >
      Direction:
    </FormLabel>
    <FormControl
      as="select"
      onChange={[Function]}
      value=""
    >
      <option
        value=""
      >
        Select a direction
      </option>
    </FormControl>
  </FormGroup>
</Form>
`;
